lib/puppet/pops/types/type_calculator.rb

```diff
@@ -452,9 +452,18 @@ def instance_of_PTupleType(t, o)
 
   def instance_of_PStructType(t, o)
     return false unless o.is_a?(Hash)
-    h = t.hashed_elements
-    # all keys must be present and have a value (even if nil/undef)
-    (o.keys - h.keys).empty? && h.all? { |k,v| instance_of(v, o[k]) }
```

Comment:
No need to change unless you want to limit the use of copies in the code. The new approach creates no hash, nor does it cache it. Only one side needs a hash to maximize efficiency during matching.